lib/utils/fs_test.go

```diff
@@ -26,6 +26,144 @@ import (
+
+	dirFilePath := filepath.Join(dirPath, "file")
+	f, err := os.Create(dirFilePath)
+	require.NoError(t, err)
+	f.Close()
+
+	dirSymlinkToFile := filepath.Join(dirPath, "file-s")
+	err = os.Symlink(dirFilePath, dirSymlinkToFile)
+	require.NoError(t, err)
+
+	symlinkDir := filepath.Join(rootDir, "dir-s")
+	err = os.Symlink(dirPath, symlinkDir)
+	require.NoError(t, err)
+
+	symlinkToSymlinkDir := filepath.Join(rootDir, "dir-s-s")
+	err = os.Symlink(symlinkDir, symlinkToSymlinkDir)
+	require.NoError(t, err)
+
+	circularSymlink := filepath.Join(rootDir, "circular-s")
+	err = os.Symlink(circularSymlink, circularSymlink)
+	require.NoError(t, err)
+
+	brokenSymlink := filepath.Join(rootDir, "broken-s")
+	err = os.Symlink(filepath.Join(rootDir, "nonexistent"), brokenSymlink)
+	require.NoError(t, err)
+
+	dirHardfilePath := filepath.Join(rootDir, "hardfile")
+	f, err = os.Create(dirHardfilePath)
+	require.NoError(t, err)
+	f.Close()
+
+	dirHardLinkToHardfile := filepath.Join(rootDir, "hardfile-h")
+	err = os.Link(dirHardfilePath, dirHardLinkToHardfile)
+	require.NoError(t, err)
+
+	// Define and run tests
+	type testCase struct {
+		name        string
+		filePath    string
+		allowSymln  bool
+		allowHardln bool
+		expectErr   string
+	}
+	testCases := []testCase{
+		{
+			name:        "directFileOpenAllowed",
+			filePath:    dirFilePath,
+			allowSymln:  false,
+			allowHardln: false,
+			expectErr:   "",
+		},
+		{
+			name:        "symlinkFileOpenAllowed",
+			filePath:    dirSymlinkToFile,
+			allowSymln:  true,
+			allowHardln: false,
+			expectErr:   "",
+		},
+		{
+			name:        "hardlinkFileOpenAllowed",
+			filePath:    dirHardLinkToHardfile,
+			allowSymln:  false,
+			allowHardln: true,
+			expectErr:   "",
+		},
+		{
+			name:        "symlinkFileOpenDenied",
+			filePath:    dirSymlinkToFile,
+			allowSymln:  false,
+			allowHardln: false,
+			expectErr:   "symlink",
+		},
+		{
+			name:        "symlinkDirFileOpenDenied",
+			filePath:    filepath.Join(symlinkDir, "file"),
+			allowSymln:  false,
+			allowHardln: false,
+			expectErr:   "symlink",
+		},
+		{
+			name:        "symlinkRecursiveDirFileOpenDenied",
+			filePath:    filepath.Join(symlinkToSymlinkDir, "file"),
+			allowSymln:  false,
+			allowHardln: false,
+			expectErr:   "symlink",
+		},
+		{
+			name:        "circularSymlink",
+			filePath:    circularSymlink,
+			allowSymln:  false,
+			allowHardln: false,
+			expectErr:   "symlink",
+		},
+		{
+			name:        "brokenSymlink",
+			filePath:    brokenSymlink,
+			allowSymln:  false,
+			allowHardln: false,
+			expectErr:   "symlink",
+		},
+		{
+			name:        "hardlinkFileOpenDenied",
+			filePath:    dirHardLinkToHardfile,
+			allowSymln:  false,
+			allowHardln: false,
+			expectErr:   "hardlink",
+		},
+	}
+
+	for _, tt := range testCases {
+		t.Run(tt.name, func(t *testing.T) {
+			f, err := openFile(tt.filePath, tt.allowSymln, tt.allowHardln)
+			if tt.expectErr == "" {
+				require.NoError(t, err)
+				f.Close()
```

Comment:
nit: Close() should happen based on the actual outcome of the method, not be based on the test expectations, as the latter may break.